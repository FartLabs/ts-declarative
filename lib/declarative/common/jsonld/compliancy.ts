import { assert } from "@std/assert/assert";
// @deno-types="npm:@types/jsonld"
import { default as jsonldjs } from "jsonld";
import type { QueryEngine } from "@comunica/query-sparql-link-traversal";
import type { Class } from "#/lib/declarative/declarative.ts";
import { getPrototypeValue } from "#/lib/declarative/declarative.ts";
import type { Context } from "#/lib/declarative/common/jsonld/context.ts";
import type { ValueJSONLd } from "#/lib/declarative/common/jsonld/jsonld.ts";
import type { ValueTsMorph } from "#/lib/declarative/common/ts-morph/ts-morph.ts";

export async function assertCompliancy(
  queryEngine: QueryEngine,
  target: Class,
) {
  const value = getPrototypeValue<ValueJSONLd & ValueTsMorph>(target);
  if (value === undefined) {
    throw new Error(
      `Class ${target.name} is missing a JSON-LD or ts-morph value.`,
    );
  }

  const typeID = (await expandKeys(value.context!, [value.type!.at(0)!])).at(
    0,
  )!;
  if (typeID === undefined) {
    throw new Error(`Type does not exist.`);
  }

  const propertyIDs = await expandKeys(
    value.context!,
    value.tsMorph!.properties.map((property) => property.name),
  );
  if (propertyIDs === undefined) {
    throw new Error(`Property does not exist.`);
  }

  // TODO: Expand keys in makeCompliancyQuery generation procedure.
  const query = makeCompliancyQuery(typeID, propertyIDs);
  const isCompliant = await queryEngine.queryBoolean(query);
  assert(isCompliant, `Class ${target.name} is not compliant.`);
}

export async function expandKeys(
  context: Context,
  keys: string[],
): Promise<string[]> {
  const expanded = await jsonldjs.expand({
    "@context": context,
    ...Object.fromEntries(keys.map((key) => [key, ""])),
  });
  return Object.keys(expanded);
}

/**
 * makeCompliancyQuery makes a SPARQL query that asserts that the given set of
 * properties are included in the given class's domain.
 */
export function makeCompliancyQuery(classID: string, propertyIDs: string[]) {
  return `ASK {
${
    propertyIDs
      .flatMap((propertyID) => {
        return [
          `<${propertyID}> <https://schema.org/domainIncludes> <${classID}> .`,
        ];
      })
      .join("\n")
  }
}`;
}
