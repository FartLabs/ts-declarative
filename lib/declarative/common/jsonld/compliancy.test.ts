import { assertSnapshot } from "@std/testing/snapshot";
// import { QueryEngine } from "@comunica/query-sparql-link-traversal";
// import { jsonSchemaDecoratorFactoryOfFile } from "#/lib/declarative/common/json-schema/json-schema-file.ts";
// import { jsonld } from "./jsonld.ts";
import { makeCompliancyQuery } from "./compliancy.ts";

// const jsonSchema = await jsonSchemaDecoratorFactoryOfFile(import.meta.url);

// @jsonld({ context: "https://schema.org/" })
// @jsonSchema()
// class Person {
//   public constructor(public name: string) {}
// }

// const queryEngine = new QueryEngine();

Deno.test("makeCompliancyQuery makes query", async (t) => {
  const query = makeCompliancyQuery("https://schema.org/Person", [
    "https://schema.org/name",
  ]);

  await assertSnapshot(t, query);
});

// TODO: Test compliancy assertion.
// Deno.test("assertCompliancy asserts class is compliant", async () => {
//   await assertCompliancy(queryEngine, Person);
// });
