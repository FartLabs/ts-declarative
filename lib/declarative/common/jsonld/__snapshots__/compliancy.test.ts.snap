export const snapshot = {};

snapshot[`makeCompliancyQuery makes query 1`] = `
"ASK {
<https://schema.org/name> <https://schema.org/domainIncludes> <https://schema.org/Person> .
}"
`;
